bot/pagination.py

```diff
@@ -0,0 +1,225 @@
+            if self._linecount >= self.max_lines:
+                self._linecount = 0
+                self.close_page()
+
+            self._linecount += 1
+        if self._count + len(line) + 1 > self.max_size:
+            self.close_page()
+
+        self._count += len(line) + 1
+        self._current_page.append(line)
+
+        if empty:
+            self._current_page.append('')
+            self._count += 1
+
+    @classmethod
+    async def paginate(cls, lines: Iterable[str], ctx: Context, embed: Embed,
+                       prefix: str = "", suffix: str = "", max_lines: Optional[int] = None, max_size: int = 500,
+                       empty: bool = True, restrict_to_user: User = None, timeout: int=300,
+                       footer_text: str = None):
+        """
+        Use a paginator and set of reactions to provide pagination over a set of lines. The reactions are used to
+        switch page, or to finish with pagination.
+        When used, this will send a message using `ctx.send()` and apply a set of reactions to it. These reactions may
+        be used to change page, or to remove pagination from the message. Pagination will also be removed automatically
+        if no reaction is added for five minutes (300 seconds).
+        >>> embed = Embed()
+        >>> embed.set_author(name="Some Operation", url=url, icon_url=icon)
+        >>> await LinePaginator.paginate(
+        ...     (line for line in lines),
+        ...     ctx, embed
+        ... )
+        :param lines: The lines to be paginated
+        :param ctx: Current context object
+        :param embed: A pre-configured embed to be used as a template for each page
+        :param prefix: Text to place before each page
+        :param suffix: Text to place after each page
+        :param max_lines: The maximum number of lines on each page
+        :param max_size: The maximum number of characters on each page
+        :param empty: Whether to place an empty line between each given line
+        :param restrict_to_user: A user to lock pagination operations to for this message, if supplied
+        :param timeout: The amount of time in seconds to disable pagination of no reaction is added
+        :param footer_text: Text to prefix the page number in the footer with
+        """
+
+        def event_check(reaction_: Reaction, user_: Member):
+            """
+            Make sure that this reaction is what we want to operate on
+            """
+
+            return (
+                reaction_.message.id == message.id and  # Reaction on this specific message
+                reaction_.emoji in PAGINATION_EMOJI and  # One of the reactions we handle
+                user_.id != ctx.bot.user.id and (  # Not applied by the bot itself
+                    not restrict_to_user or   # Unrestricted if there's no user to restrict to, or...
+                    user_.id == restrict_to_user.id  # Only by the restricted user
+                )
```

Comment:
Could we have something a bit more readable for the conditions? Maybe something like:
```py
no_restrictions = (
    # Pagination is not restricted
    not restrict_to_user or
    # The reaction was by a whitelisted user
    user_.id == restrict_to_user.id
)

return (
    # Conditions for a successful pagination:
    all(
        # Reaction is on this message
        reaction_.message.id == message.id,
        # Reaction is one of the pagination emotes
        reaction_.emoji in PAGINATION_EMOJI,
        # Reaction was not made by the Bot
        user_.id != ctx.bot.user.id,
        # There were no restrictions
        no_restrictions
    )
)
```